packages/@n8n/ai-workflow-builder.ee/src/ai-workflow-builder-agent.service.ts

```diff
@@ -124,51 +123,33 @@ export class AiWorkflowBuilderService {
 		}
 	}
 
-	private getNodeTypes(): INodeTypeDescription[] {
-		// These types are ignored because they tend to cause issues when generating workflows
-		const ignoredTypes = [
+	private filterNodeTypes(nodeTypes: INodeTypeDescription[]): INodeTypeDescription[] {
+		const ignoredTypes = new Set([
 			'@n8n/n8n-nodes-langchain.toolVectorStore',
 			'@n8n/n8n-nodes-langchain.documentGithubLoader',
 			'@n8n/n8n-nodes-langchain.code',
-		];
-		const nodeTypesKeys = Object.keys(this.nodeTypes.getKnownTypes());
-
-		const nodeTypes = nodeTypesKeys
-			.filter((nodeType) => !ignoredTypes.includes(nodeType))
-			.map((nodeName) => {
-				try {
-					return { ...this.nodeTypes.getByNameAndVersion(nodeName).description, name: nodeName };
-				} catch (error) {
-					this.logger?.error('Error getting node type', {
-						nodeName,
-						error: error instanceof Error ? error.message : 'Unknown error',
-					});
-					return undefined;
-				}
-			})
-			.filter(
-				(nodeType): nodeType is INodeTypeDescription =>
-					// We filter out hidden nodes, except for the Data Table node which has custom hiding logic
-					// See more details in DataTable.node.ts#L29
-					nodeType !== undefined &&
-					(nodeType.hidden !== true || nodeType.name === 'n8n-nodes-base.dataTable'),
-			)
-			.map((nodeType, _index, nodeTypes: INodeTypeDescription[]) => {
-				// If the node type is a tool, we need to find the corresponding non-tool node type
-				// and merge the two node types to get the full node type description.
-				const isTool = nodeType.name.endsWith('Tool');
-				if (!isTool) return nodeType;
-
-				const nonToolNode = nodeTypes.find((nt) => nt.name === nodeType.name.replace('Tool', ''));
-				if (!nonToolNode) return nodeType;
-
-				return {
-					...nonToolNode,
-					...nodeType,
-				};
-			});
+		]);
+
+		const visibleNodeTypes = nodeTypes.filter(
+			(nodeType) =>
+				!ignoredTypes.has(nodeType.name) &&
+				(nodeType.hidden !== true || nodeType.name === 'n8n-nodes-base.dataTable'),
```

Comment:
Might be worth keeping the comment explaining that the data table is considered hidden but has its own custom hiding logic